pupil_src/shared_modules/eye_movement_detector.py

```diff
@@ -0,0 +1,937 @@
+"""
+(*)~---------------------------------------------------------------------------
+Pupil - eye tracking platform
+Copyright (C) 2012-2019 Pupil Labs
+
+Distributed under the terms of the GNU
+Lesser General Public License (LGPL v3.0).
+See COPYING and COPYING.LESSER for license details.
+---------------------------------------------------------------------------~(*)
+"""
+
+
+# stdlib
+import os
+import csv
+import enum
+import typing
+import logging
+import traceback
+from collections import deque
+from typing import _Protocol as Protocol
+
+
+# local
+from tasklib import interface
+from tasklib.background.task import BackgroundGeneratorFunction
+from tasklib.background.patches import Patch, IPCLoggingPatch
+from tasklib.manager import PluginTaskManager
+import file_methods as fm
+import player_methods as pm
+
+import methods
+from plugin import Analysis_Plugin_Base
+from observable import Observable
+import video_capture as vc
+
+# third-party
+import nslr_hmm
+import numpy as np
+import cv2
+from pyglui import ui
+from pyglui.cygl.utils import RGBA, draw_circle
+from pyglui.pyfontstash import fontstash
+
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: This protocol definition can be moved into `camera_models.py`
+class Camera_Model(Protocol):
+    def undistort(self, img: np.ndarray) -> np.ndarray:
+        ...
+
+    def unprojectPoints(
+        self, pts_2d: np.ndarray, use_distortion: bool = True, normalize: bool = False
+    ) -> np.ndarray:
+        ...
+
+    def projectPoints(
+        self,
+        object_points,
+        rvec: typing.Optional[np.ndarray] = None,
+        tvec: typing.Optional[np.ndarray] = None,
+        use_distortion: bool = True,
+    ):
+        ...
+
+    # def solvePnP(self, uv3d, xy): ...
+    def save(self, directory: str, custom_name: typing.Optional[str] = None):
+        ...
+
+
+class Immutable_Capture:
+    def __init__(self, capture: typing.Type[vc.base_backend.Base_Source]):
+        self.frame_size: typing.Tuple[int, int] = (
+            int(capture.frame_size[0]),
+            int(capture.frame_size[1]),
+        )
+        self.intrinsics: Camera_Model = capture.intrinsics
+        self.timestamps: np.ndarray = capture.timestamps
+
+
+Color_RGBA = typing.Tuple[float, float, float, float]
+
+
+@enum.unique
+class Segment_Method(enum.Enum):
+    GAZE = "gaze"
+    PUPIL = "pupil"
+
+
+@enum.unique
+class Segment_Class(enum.Enum):
+    FIXATION = "fixation"
+    SACCADE = "saccade"
+    POST_SACCADIC_OSCILLATIONS = "pso"
+    SMOOTH_PURSUIT = "smooth_pursuit"
+
+    @classmethod
+    def from_nslr_class(cls, nslr_class):
+        return {
+            nslr_hmm.FIXATION: Segment_Class.FIXATION,
+            nslr_hmm.SACCADE: Segment_Class.SACCADE,
+            nslr_hmm.PSO: Segment_Class.POST_SACCADIC_OSCILLATIONS,
+            nslr_hmm.SMOOTH_PURSUIT: Segment_Class.SMOOTH_PURSUIT,
+        }[nslr_class]
```

Comment:
How about making this a class attribute. This would change the calling syntax only slightly different and the dict would be instantiated only once.